sel4-tracebuffer.c

```diff
@@ -0,0 +1,416 @@
+	.owner   = THIS_MODULE,
+	.open    = sel4_tracebuffer_seq_open,
+	.read    = seq_read,
+	.llseek  = seq_lseek,
+	.release = sel4_tracebuffer_seq_release,
+};
+
+static const struct file_operations trace_on_fops = {
+	.owner   = THIS_MODULE,
+	.open    = sel4_open,
+	.read    = sel4_trace_on_read,
+	.write   = sel4_trace_on_write,
+	.release = sel4_release,
+};
+
+static struct dentry *sel4_tracebuffer_dir;
+
+void sel4_tracebuffer_register_sysfs(void)
+{
+	sel4_tracebuffer_dir = debugfs_create_dir("sel4_tracebuffer", NULL);
+
+	/*
+	 * trace -- txt data
+	 */
+	debugfs_create_file("trace", 0444, sel4_tracebuffer_dir,
+			    &sel4_tracebuffer_seq_ops, &trace_fops);
+	/*
+	 * tracedata -- binary data
+	 */
+	debugfs_create_file("tracedata", 0444, sel4_tracebuffer_dir,
+			    &sel4_trace_data_seq_ops, &trace_data_fops);
+
+	/*
+	 * trace_on -- trace on (1) / trace off (0)
+	 */
+	debugfs_create_file("trace_on", 0644, sel4_tracebuffer_dir,
+			    NULL, &trace_on_fops);
+
+	mutex_init(&pdata_local.mutex);
+}
+
+void sel4_tracebuffer_unregister_sysfs(void)
+{
+	debugfs_remove_recursive(sel4_tracebuffer_dir);
+}
+
+static void *sel4_tracebuffer_vmap(phys_addr_t start, size_t size,
+		unsigned int memtype)
+{
+	struct page **pages;
+	phys_addr_t page_start;
+	unsigned int page_count;
+	pgprot_t prot;
+	unsigned int i;
+	void *vaddr;
+
+	page_start = start - offset_in_page(start);
+	page_count = DIV_ROUND_UP(size + offset_in_page(start), PAGE_SIZE);
+
+	if (memtype)
+		prot = pgprot_noncached(PAGE_KERNEL);
+	else
+		prot = pgprot_writecombine(PAGE_KERNEL);
+
+	pages = kmalloc_array(page_count, sizeof(struct page *), GFP_KERNEL);
+
+	if (!pages)
+		return NULL;
+
+	for (i = 0; i < page_count; i++) {
+		phys_addr_t addr = page_start + i * PAGE_SIZE;
+
+		pages[i] = pfn_to_page(addr >> PAGE_SHIFT);
+	}
+	vaddr = vmap(pages, page_count, VM_MAP, prot);
+	kfree(pages);
+
+	/*
+	 * Since vmap() uses page granularity, we must add the offset
+	 * into the page here, to get the byte granularity address
+	 * into the mapping to represent the actual "start" location.
+	 */
+	return vaddr + offset_in_page(start);
+}
+
+static void *sel4_tracebuffer_iomap(phys_addr_t start, size_t size,
+		unsigned int memtype, char *label)
+{
+	void *va;
+
+	if (!request_mem_region(start, size, label ?: "ramoops")) {
+		pr_err("%s: request mem region (%s 0x%llx@0x%llx) failed\n",
+			__func__, label ?: "ramoops",
+			(unsigned long long)size, (unsigned long long)start);
+		return NULL;
+	}
+
+	if (memtype)
+		va = ioremap(start, size);
+	else
+		va = ioremap_wc(start, size);
+
+	/*
+	 * Since request_mem_region() and ioremap() are byte-granularity
+	 * there is no need handle anything special like we do when the
+	 * vmap() case in sel4_tracebuffer_ram_vmap() above.
+	 */
+	return va;
+}
+
+static void *sel4_tracebuffer_map(phys_addr_t start, phys_addr_t size, int memtype)
+{
+	void *vaddr = NULL;
+
+	if (pfn_valid(start >> PAGE_SHIFT))
```

Comment:
Again I would say that we have 2 options here: keep code as is or remove that code and place here error in case someone somewhere in future will try so. Both options are fine for me. What do you suggest here?

Reply:
As agreed separate I put here a comment and also a WARNING notification.